platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/DefaultConfigurationCache.kt

```diff
@@ -410,79 +413,105 @@ class DefaultConfigurationCache internal constructor(
 
         if (startParameter.isIgnoreInputsInTaskGraphSerialization) {
             InstrumentedInputs.discardListener()
         }
 
-        buildOperationRunner.withStoreOperation(cacheKey.string) {
-            val stateStoreResult = store.useForStore { layout ->
-                try {
-                    val stateFile = layout.fileFor(stateType)
-                    action(stateFile)
-                    val storeFailure = problems.queryFailure()
-                    storeFailure
-                } catch (error: Exception) {
-                    // Invalidate state on serialization errors
-                    problems.failingBuildDueToSerializationError()
-                    throw error
-                }
+        buildOperationRunner.withWorkGraphStoreOperation(cacheKey.string) {
+            val stateStoreResult = runAndStore(stateType = StateType.Work) { stateFile: ConfigurationCacheStateFile ->
+                writeConfigurationCacheState(stateFile)
             }
-
-            val storeFailure = stateStoreResult.value
-            StoreResult(stateStoreResult.accessedFiles, storeFailure)
+            WorkGraphStoreResult(stateStoreResult.accessedFiles, stateStoreResult.value)
         }
 
         crossConfigurationTimeBarrier()
     }
 
     private
-    data class LoadResultMetadata(val originInvocationId: String? = null)
-
-    private
-    fun loadModel(): Any {
-        return loadFromCache(StateType.Model) { stateFile ->
-            LoadResultMetadata() to cacheIO.readModelFrom(stateFile)
+    fun runAndStore(
+        stateType: StateType,
+        action: (ConfigurationCacheStateFile) -> Unit
+    ): ConfigurationCacheStateStore.StateAccessResult<Throwable?> {
+
+        return store.useForStore { layout ->
+            try {
+                val stateFile = layout.fileFor(stateType)
+                action(stateFile)
+                val storeFailure = problems.queryFailure()
+                storeFailure
+            } catch (error: Exception) {
+                // Invalidate state on serialization errors
+                problems.failingBuildDueToSerializationError()
+                throw error
+            }
         }
     }
 
     private
-    fun loadWorkGraph(graph: BuildTreeWorkGraph, graphBuilder: BuildTreeWorkGraphBuilder?, loadAfterStore: Boolean): BuildTreeWorkGraph.FinalizedGraph {
-        return loadFromCache(StateType.Work) { stateFile ->
-            val (buildInvocationId, workGraph) = cacheIO.readRootBuildStateFrom(stateFile, loadAfterStore, graph, graphBuilder)
-            LoadResultMetadata(buildInvocationId) to workGraph
+    data class LoadResultMetadata(val originInvocationId: String? = null)
+
+    private
+    fun loadModel(): Any = runAtConfigurationTime {
+        // TODO:configuration-cache is this appliccable to model storing as well?
+        // No need to record the `ClassLoaderScope` tree
+        // when loading the task graph.
+        scopeRegistryListener.dispose()
```

Comment:
❓  Is this applicable to the model loading?

This was also part of the shared helper function, which might have unintentionally affected model loading in a non-trivial way.

Reply:
If model creation (and serialization) can still happen after a `loadModel` call, then the `scopeRegistryListener` must not be disposed, otherwise, classloaders from scopes created after this call will no longer be recognized by the CC infrastructure.

Reply:
In the current implementation `loadModel` is a "final" operation for a build action, so no serialization should happen afterward. So I'll keep this early disposal here as well.

Reply:
That makes sense.

Perhaps it would be good to introduce an assertion somewhere that avoids CC serialization after `scopeRegistryListener` has been disposed of.